storage-provider/server/src/storage.rs

```diff
@@ -194,20 +198,27 @@ async fn upload(
 
     // Branching needed here since the resulting `StreamReader`s don't have the same type
     let file_cid = if request.headers().contains_key("Content-Type") {
-        // Handle multipart forms
-        let mut multipart = Multipart::from_request(request, &s)
+        // Handle the multipart data
+        let mut multipart = Multipart::from_request(request, &state)
             .await
             .map_err(|err| (StatusCode::BAD_REQUEST, err.to_string()))?;
-        let Some(field) = multipart
+
+        // Get the field and read it entirely into memory
+        let field = multipart
             .next_field()
-            .map_err(|err| (StatusCode::BAD_REQUEST, err.to_string()))
-            .await?
-        else {
-            return Err((StatusCode::BAD_REQUEST, "empty request".to_string()));
-        };
+            .await
+            .map_err(|err| (StatusCode::BAD_REQUEST, err.to_string()))?
+            .ok_or((StatusCode::BAD_REQUEST, "empty request".to_string()))?;
```

Comment:
You're forcing a string allocation without need
```suggestion
            .ok_or_else(|| (StatusCode::BAD_REQUEST, "empty request".to_string()))?;
```